core/src/main/java/tech/pegasys/eth2signer/core/Runner.java

```diff
@@ -127,22 +139,74 @@ public void run() {
       metricsEndpoint.start(vertx);
 
       final AzureKeyVaultSignerFactory azureFactory = new AzureKeyVaultSignerFactory();
-
-      final ArtifactSignerProviderFactory factory =
-          new ArtifactSignerProviderFactory(metricsSystem, vertx, azureFactory);
-
-      final ArtifactSignerProvider blsSignerProvider =
-          factory.createBlsSignerProvider(config.getKeyConfigPath());
-
-      final ArtifactSignerProvider ethSecpSignerProvider =
-          factory.createEthSecpSignerProvider(config.getKeyConfigPath());
-
-      final ArtifactSignerProvider fcSecpSignerProvider =
-          factory.createFilecoinSecpSignerProvider(
-              config.getKeyConfigPath(), config.getFilecoinNetwork());
+      final HashicorpConnectionFactory hashicorpConnectionFactory =
+          new HashicorpConnectionFactory(vertx);
+
+      final BlsArtifactSignerFactory blsArtifactSignerFactory =
+          new BlsArtifactSignerFactory(
+              config.getKeyConfigPath(), metricsSystem, hashicorpConnectionFactory);
+
+      final Secp256k1ArtifactSignerFactory ethSecpArtifactSignerFactory =
+          new Secp256k1ArtifactSignerFactory(
+              hashicorpConnectionFactory,
+              config.getKeyConfigPath(),
+              azureFactory,
+              EthSecpArtifactSigner::new);
+
+      final Secp256k1ArtifactSignerFactory fcSecpArtifactSignerFactory =
+          new Secp256k1ArtifactSignerFactory(
+              hashicorpConnectionFactory,
+              config.getKeyConfigPath(),
+              azureFactory,
+              signer -> new FcSecpArtifactSigner(signer, config.getFilecoinNetwork()));
+
+      final Collection<ArtifactSigner> signers =
+          SignerLoader.load(
+              config.getKeyConfigPath(),
+              "yaml",
+              new YamlSignerParser(
+                  blsArtifactSignerFactory,
+                  ethSecpArtifactSignerFactory,
+                  fcSecpArtifactSignerFactory));
+
+      final Map<SignerTypes, List<ArtifactSigner>> signersByType =
+          signers
+              .parallelStream()
+              .collect(
+                  Collectors.groupingBy(
+                      i -> {
+                        if (i instanceof BlsArtifactSigner) {
```

Comment:
Should a ArtifactSigner simply have an associated SignerType? Then we could just do groupingBy(ArtifactSigner::getSignerType)

Reply:
Yeah, I was thinking the same ... but equally, it only fixes this one bit of yuck, which is pretty well controlled (and only invoked once).